### PR TITLE
Fix #10489 Problems with GeoStory map configurations merge process 

### LINKED ISSUE #10235: Interactive legend in dashboard and geostory
## Description

We need to add the interactive legend behavior also in dashboard.
Expected implementations:

- include the layerFilter among the properties updated by the legend widget (Note: probably it's enough to pass the layerFilter [here](https://github.com/geosolutions-it/MapStore2/blob/d8243f08b6e0e4bdfc1e3119992cd79196e536bd/web/client/components/widgets/enhancers/legendWidget.js#L57-L62))
- fix the wrong behaviour of the TOC for map widgets where the legend is restored to default configuration (see video)

https://github.com/geosolutions-it/MapStore2/assets/19175505/7a519709-97ff-493c-9abd-971f412af22e

It is also required to  add the interactive legend behavior for geostory as well 

**What kind of improvement you want to add?** (check one with "x", remove the others)
- [x] Minor changes to existing features

### LINKED ISSUE #10489: Problems with GeoStory map configurations merge process
## Description

Legacy geostories used following configuration to store the difference between original resources and changes made to a map in a specific section:

original map resource
```
{
    "groups": [
        {
            "id": "group-02",
            "title": "over",
            "expanded": false
        },
        {
            "id": "group-01",
            "title": "base",
            "expanded": true
        }
    ],
    "layers": [
        {
            "id": "layer-01",
            "group": "group-01",
            "source": "osm",
            "name": "mapnik",
            "title": "Open Street Map",
            "type": "osm",
            "visibility": true
        },
        {
            "id": "layer-02",
            "format": "image/jpeg",
            "group": "group-02",
            "name": "workspace:layer",
            "opacity": 1,
            "title": "Layer",
            "type": "wms",
            "url": "/geoserver/wms",
            "visibility": false
        }
    ]
}
```

diff section map configuration
```js
{
    "groups": [
        null,
        {
            "expanded": false
        }
    ],
    "layers": [
        null,
        {
            "visibility": true
        }
    ]
}
```

The merging process of these two configuration has two steps:

- the diff configuration [was cleaned](https://github.com/geosolutions-it/MapStore2/blob/v2023.02.02/web/client/components/geostory/common/enhancers/map.jsx#L40) by replacing null layers with undefined
- then the [createMapObject](https://github.com/geosolutions-it/MapStore2/blob/v2023.02.02/web/client/utils/GeoStoryUtils.js#L176-L178) method was merging the two configuration so layers and groups array items were merged by index

This process had a problem in the step 1 because only null layers were converted but not group layers.

With the introduction of the new TOC the diff configuration for layers and groups was removed and this two field are cloned as soon there is a modification in the section. This introduced another problem described in this comment:

https://github.com/geosolutions-it/MapStore2/pull/10426#issuecomment-2176031801 

In conclusin we need to solve the issues described above:

- [ ] legacy geostory diff configuration should work also when groups are null
- [ ] new geostory diff configuration should not use the merge method for layers and groups becouse this two field are cloned using the new TOC

## How to reproduce

-  import this legacy geostory on MapStore: [legacy-geostory-with-null-groups.zip](https://github.com/user-attachments/files/16360425/legacy-geostory-with-null-groups.zip)

*Expected Result*

The geostory should be visible and working

*Current Result*

The application show the error message on full screen

- [x] Not browser related

<details><summary> <b>Browser info</b> </summary>

(use this site: <a href="https://www.whatsmybrowser.org/">https://www.whatsmybrowser.org/</a> for non expert users)

| Browser Affected | Version |
|---|---|
|Internet Explorer| |
|Edge| |
|Chrome| |
|Firefox| |
|Safari| |
</details>

### PR DESCRIPTION
**Pull request by mahmoudadel54**
## Description
In this PR, fixing issue of reset legend filter within map configure panel in geostory is implemented. Now in geostories, if the interactive legend is enabled and set on a new map, it is possible to edit it correctly on the map inline panel.
The issue is mentioned here: https://github.com/geosolutions-it/MapStore2/pull/10426#issuecomment-2176031801

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [x] Bugfix
 - [ ] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

## Issue
https://github.com/geosolutions-it/MapStore2/pull/10426#issuecomment-2176031801
Also involved https://github.com/geosolutions-it/MapStore2/issues/10489

**What is the current behavior?**
https://github.com/geosolutions-it/MapStore2/pull/10426#issuecomment-2176031801

**What is the new behavior?**
In geostories, if the interactive legend is enabled and set on a new map, it is Ok to edit it correctly on the map inline panel without any non expected behavior.

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

### CODE REVIEW

File: web/client/utils/__tests__/GeoStoryUtils-test.js [hunk 1]
```diff
@@ -355,8 +355,11 @@ describe("GeoStory Utils", () => {
         };
         const res = createMapObject(DEFAULT_MAP_OPTIONS, {
             mapOptions: {
+                scrollWheelZoom: false,
                 interactions: {
-                    mouseClick: false
+                    mouseWheelZoom: false,
+                    mouseClick: false,
+                    dragPan: true
                 }
```
Comment (allyoucanmap) [inline web/client/utils/__tests__/GeoStoryUtils-test.js:363]: I would avoid to change the test structure.
Here it seems that the expectation is to merge these value so probably we should apply the replacement just for the layer key.

We will need also an additional test for the layers key

File: web/client/utils/GeoStoryUtils.js [hunk 2]
```diff
@@ -174,7 +174,13 @@ export const applyDefaults = (options = {}) => merge({}, DEFAULT_MAP_OPTIONS, op
...
  */
 export const createMapObject = (baseMap = {}, overrides = {}) => {
-    return merge({}, baseMap, overrides);
+    const baseMapCloned = { ...baseMap };
+    // use override [replace] logic instead of merge to get the updated map
+    const updatedMap = Object.keys(overrides).reduce((cumulatedMap, key) => {
+        cumulatedMap[key] = overrides[key];
+        return cumulatedMap;
+    }, baseMapCloned);
+    return updatedMap;
```
Comment (allyoucanmap) [inline web/client/utils/GeoStoryUtils.js:183]: Based on the previous comment I think we should merge all the properties but the layers key that will be overridden

**Review by allyoucanmap — CHANGES_REQUESTED** [review]

**Review by allyoucanmap — CHANGES_REQUESTED** [review]
@tdipisa @mahmoudadel54 This fix should take into account also the problem described in this other issue https://github.com/geosolutions-it/MapStore2/issues/10489.
I'm going to change the title of PR and connect it to the new issue

**Comment by tdipisa** [comment]
> @tdipisa @mahmoudadel54 This fix should take into account also the problem described in this other issue #10489. I'm going to change the title of PR and connect it to the new issue

@mahmoudadel54 can you please proceed as requested above providing an estimate before?

**Review by allyoucanmap — APPROVED** [review]

**Comment by allyoucanmap** [comment]
@ElenaGallo please test this fix on dev, thanks

**Pull request merged**